### PR TITLE
Bugfix/fix prefix of meta.json file

### LINKED ISSUE #2585: Tempo failed to find `meta.json` path after adding prefix in S3 configuration.
**Describe the bug**
Tempo failed to find the meta.json file path after adding a prefix in S3 configuration.
We were watching issue #2100 which is what we desired to have. 
This `prefix` support was not in 2.1.1 release, so we build the Tempo from the source code.
And we found the blocks can be saved as expected with the `prefix`, but compaction and deletion of these blocks failed.

_with prefix_
> level=info ts=2023-06-21T16:49:25.6918833Z caller=retention.go:68 msg="marking block for deletion" blockID=a5916bb1-9132-4340-8410-4f4f56e50643 tenantID=single-tenant
level=error ts=2023-06-21T16:49:25.6945432Z caller=retention.go:71 msg="failed to mark block compacted during retention" blockID=a5916bb1-9132-4340-8410-4f4f56e50643 tenantID=single-tenant err="error copying obj meta to compacted obj meta: The specified key does not exist."
level=info ts=2023-06-21T16:49:25.7054214Z caller=poller.go:209 msg="writing tenant index" tenant=single-tenant metas=2 compactedMetas=0
level=info ts=2023-06-21T16:49:25.7059633Z caller=poller.go:209 msg="writing tenant index" tenant=single-tenant metas=2 compactedMetas=0
level=info ts=2023-06-21T16:49:25.7124402Z caller=poller.go:129 msg="blocklist poll complete" seconds=0.0205569
level=info ts=2023-06-21T16:49:25.7425927Z caller=poller.go:129 msg="blocklist poll complete" seconds=0.053963

with-prefix-block
As shown there was no `meta.compacted.json`

_without prefix_

> level=info ts=2023-06-21T17:14:16.8284437Z caller=retention.go:68 msg="marking block for deletion" blockID=7ac206e8-a881-409e-956c-efa8b878adc8 tenantID=single-tenant
level=info ts=2023-06-21T17:14:16.8361876Z caller=poller.go:209 msg="writing tenant index" tenant=single-tenant metas=2 compactedMetas=1
level=info ts=2023-06-21T17:14:16.8361876Z caller=poller.go:209 msg="writing tenant index" tenant=single-tenant metas=2 compactedMetas=1
level=info ts=2023-06-21T17:14:16.8382732Z caller=retention.go:97 msg="deleting block" blockID=33f983f0-76ba-4564-9a7b-e8fa6e08008b tenantID=single-tenant
level=info ts=2023-06-21T17:14:16.8438316Z caller=poller.go:129 msg="blocklist poll complete" seconds=0.0153879
level=info ts=2023-06-21T17:14:16.8790127Z caller=poller.go:129 msg="blocklist poll complete" seconds=0.0505142

without-prefix-block

Remove the `prefix` from the configuration it seems everything is back to normal.

**To Reproduce**
1. download source code and build tempo from the `main` branch (Jun 23, 2023)
2. Starting with docker s3 example [https://github.com/grafana/tempo/blob/main/example/docker-compose/s3/docker-compose.yaml](url) instead of the latest tempo image, using the local build tempo.
3. setting the `prefix` in storage s3 configuration

**Expected behavior**
enable `prefix` should not affect the compaction and deletion process.

**Environment:**
 - Infrastructure: Openshift, laptop

**Additional Context**
I did some experiments on my local laptop, and this issue seems to be caused by the `MetaFileName`https://github.com/grafana/tempo/blob/c63338744d3842b54240e39a049ce688e910edb8/tempodb/backend/raw.go#L238, `CompactedMetaFileName`https://github.com/grafana/tempo/blob/c63338744d3842b54240e39a049ce688e910edb8/tempodb/backend/raw.go#L243 and `RootPath`https://github.com/grafana/tempo/blob/c63338744d3842b54240e39a049ce688e910edb8/tempodb/backend/raw.go#L248, these functions do not return the file name with the `rw.cfg.Prefix`

After some adjustments in these functions and `compactor.go`, tempo seems to work correctly. 

_after adjustments_
> level=info ts=2023-06-21T19:59:37.3055968Z caller=retention.go:68 msg="marking block for deletion" blockID=9ef75b3f-4dbd-4422-b3a1-95f259ad4333 tenantID=single-tenant
level=info ts=2023-06-21T19:59:37.3193817Z caller=poller.go:209 msg="writing tenant index" tenant=single-tenant metas=2 compactedMetas=1
level=info ts=2023-06-21T19:59:37.3199077Z caller=poller.go:209 msg="writing tenant index" tenant=single-tenant metas=2 compactedMetas=1
level=info ts=2023-06-21T19:59:37.3206734Z caller=retention.go:97 msg="deleting block" blockID=b4d20bb0-d4f8-418a-ba83-07be7d2337ea tenantID=single-tenant
level=info ts=2023-06-21T19:59:37.3258373Z caller=poller.go:129 msg="blocklist poll complete" seconds=0.0202405
level=info ts=2023-06-21T19:59:37.3634428Z caller=poller.go:129 msg="blocklist poll complete" seconds=0.057846
fixed-block

### PR DESCRIPTION
**Pull request by WildCatFish**
**What this PR does**:
fix the incorrect path of `meta.json` file when enabling `prefix` in storage configuration, this issue affects compaction and deletion of blocks.
**Which issue(s) this PR fixes**:
Fixes #2585

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: tempodb/backend/raw.go [hunk 1]
```diff
@@ -235,16 +235,16 @@ func KeyPathWithPrefix(keypath KeyPath, prefix string) KeyPath {
...
-func CompactedMetaFileName(blockID uuid.UUID, tenantID string) string {
-	return path.Join(RootPath(blockID, tenantID), CompactedMetaName)
+func CompactedMetaFileName(blockID uuid.UUID, tenantID string, prefix string) string {
+	return path.Join(prefix, tenantID, blockID.String(), CompactedMetaName)
 }
 
 // RootPath returns the root path for a block given a block id and tenantid
-func RootPath(blockID uuid.UUID, tenantID string) string {
-	return path.Join(tenantID, blockID.String())
+func RootPath(blockID uuid.UUID, tenantID string, prefix string) string {
```
Comment (joe-elliott) [inline tempodb/backend/raw.go:248]: you made a change to this method but then didn't use it? should we continue to use this method above?

File: tempodb/backend/raw.go [hunk 2]
```diff
@@ -235,16 +235,16 @@ func KeyPathWithPrefix(keypath KeyPath, prefix string) KeyPath {
 }
 
 // MetaFileName returns the object name for the block meta given a block id and tenantid
-func MetaFileName(blockID uuid.UUID, tenantID string) string {
-	return path.Join(RootPath(blockID, tenantID), MetaName)
+func MetaFileName(blockID uuid.UUID, tenantID string, prefix string) string {
```
Comment (joe-elliott) [inline tempodb/backend/raw.go:238]: what happens if prefix is empty. does this do what is expected? maybe add some quick tests?

**Review by joe-elliott — COMMENTED** [review]
thx for fixing this! looks like there's some linting issues.

File: tempodb/backend/raw.go (hunk 1 above)
Comment (WildCatFish) [reply tempodb/backend/raw.go:248]: hi, this method is used by the compactor. https://github.com/grafana/tempo/blob/deeb2636b88bbbc16e288b4c5f744e50bffe20c2/tempodb/backend/s3/compactor.go#L51
this method was also used by the previous `MetaFileName` and `CompactedMetaFileName`
https://github.com/grafana/tempo/blob/f3e3c9a70b370cf1a936451cf2dbcb2404136412/tempodb/backend/raw.go#L244

File: tempodb/backend/raw.go (hunk 2 above)
Comment (WildCatFish) [reply tempodb/backend/raw.go:238]: tests are at `tempodb/backend/raw_test.go` check the commit [add tests for modifed metaFile/Path methods](https://github.com/grafana/tempo/pull/2591/commits/3051385bd0f91c18ecfd17fac65dd94bae83b36a)

**Review by joe-elliott — COMMENTED** [review]
Thanks for the answers! I am good on this, but you need to appease the linter.

File: CHANGELOG.md [hunk 3]
```diff
@@ -54,6 +54,7 @@ To make use of filtering, configure `autocomplete_filtering_enabled`.
 * [BUGFIX] metrics-generator: ensure Prometheus will scale up shards when remote write is lagging behind [#2463](https://github.com/grafana/tempo/issues/2463) (@kvrhdn)
 * [BUGFIX] Fixes issue where matches and other spanset level attributes were not persisted to the TraceQL results. [#2490](https://github.com/grafana/tempo/pull/2490) 
 * [BUGFIX] Fixes issue where ingester search could occasionally fail with file does not exist error [#2534](https://github.com/grafana/tempo/issues/2534) (@mdisibio)
+* [BUGFIX] Tempo failed to find meta.json path after adding prefix in S3 configuration. [#2585](https://github.com/grafana/tempo/issues/2585) (@WildCatFish)
```
Comment (mapno) [inline CHANGELOG.md:57]:
```suggestion
* [BUGFIX] Tempo failed to find meta.json path after adding prefix in S3/GCS/Azure configuration. [#2585](https://github.com/grafana/tempo/issues/2585) (@WildCatFish)
```

File: tempodb/backend/raw_test.go [hunk 4]
```diff
@@ -10,6 +10,11 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
+const (
+	TenantID      = "test"
+	StoragePrefix = "test/prefix"
```
Comment (mapno) [inline tempodb/backend/raw_test.go:15]:
```suggestion
	tenantID      = "test"
	storagePrefix = "test/prefix"
```

I don't think we should export these constants

**Review by mapno — COMMENTED** [review]
LGTM, just two minor comments. Thanks for the fix!

**Review by joe-elliott — APPROVED** [review]
Thanks for the fix!

**Pull request merged**